packages/debug/src/browser/debug-session.tsx

```diff
@@ -73,8 +73,12 @@ export class DebugSession implements CompositeTreeElement {
         protected readonly breakpoints: BreakpointManager,
         protected readonly labelProvider: LabelProvider,
         protected readonly messages: MessageClient,
-        protected readonly fileService: FileService) {
+        protected readonly fileService: FileService,
+        protected readonly debugCP: ContributionProvider<DebugContribution>
+    ) {
         this.connection.onRequest('runInTerminal', (request: DebugProtocol.RunInTerminalRequest) => this.runInTerminal(request));
+        this.registerDebugContributions(options.configuration.type, this.connection);
```

Comment:
So every time a new `DebugSession` is created, all contributions will be registered again? I would have expected registration to be a one time thing?

Reply:
There are 2 different parts happening here:
1. "Registration" of contributions - as you called it - are contribution providers/containers populated when app loads - once - see **DebugMainImpl** - packages\plugin-ext\src\main\browser\debug\debug-main.ts.

2. **registerDebugContributions** method here is going over list of contributions and invoking `register` method on contribution itself in order for contribution itself to register to requests/events sent from `vscode-js-debug`.  This method is not populating contributions again.